sw/airborne/modules/dragspeed/dragspeed.c

```diff
@@ -0,0 +1,262 @@
+ * - The size and weight of the drone are constant (changes in size and weight
+ *   require re-calibration).
+ * - Pitch and roll angles are small (< ~20 deg).
+ * - No wind (when used as ground speed).
+ */
+
+#include "modules/dragspeed/dragspeed.h"
+
+#include "subsystems/abi.h"
+#include "subsystems/abi_common.h"
+#include "subsystems/datalink/telemetry.h"
+
+#include <stdio.h>
+
+#ifndef DRAGSPEED_SEND_ABI_MESSAGE
+#define DRAGSPEED_SEND_ABI_MESSAGE TRUE
+#endif
+
+#ifndef DRAGSPEED_ACCEL_ID
+#define DRAGSPEED_ACCEL_ID ABI_BROADCAST
+#endif
+
+#ifndef DRAGSPEED_COEFF_X
+#define DRAGSPEED_COEFF_X 1.0 /// Drag coefficient (mu/m) of the linear drag model along x axis, where m*a = v*mu
+#endif
+
+#ifndef DRAGSPEED_COEFF_Y
+#define DRAGSPEED_COEFF_Y 1.0 /// Drag coefficient (mu/m) of the linear drag model along y axis, where m*a = v*mu
+#endif
+
+#ifndef DRAGSPEED_R
+#define DRAGSPEED_R 0.25 /// Velocity measurement noise variance [(m/s)^2]
+#endif
+
+#ifndef DRAGSPEED_FILTER
+#define DRAGSPEED_FILTER 0.8 /// First-order low-pass filter strength [0..1]. Pretty high default value as accelero's tend to be noisy.
+#endif
+
+struct dragspeed_t dragspeed;
+
+static abi_event accel_ev;
+static void accel_cb(uint8_t sender_id, uint32_t stamp,
+    struct Int32Vect3 *accel);
+
+static void send_dragspeed(struct transport_tx *trans, struct link_device *dev);
+
+static void calibrate_coeff(struct Int32Vect3 *accel);
+static void calibrate_zero(struct Int32Vect3 *accel);
+
+void dragspeed_init(void)
+{
+  // Set initial values
+  dragspeed.coeff.x = DRAGSPEED_COEFF_X;
+  dragspeed.coeff.y = DRAGSPEED_COEFF_Y;
+  dragspeed.filter = DRAGSPEED_FILTER;
+#ifdef DRAGSPEED_ZERO_X
+  dragspeed.zero.x = DRAGSPEED_ZERO_X;
+#endif
+#ifdef DRAGSPEED_ZERO_Y
+  dragspeed.zero.y = DRAGSPEED_ZERO_Y;
+#endif
+#if defined(DRAGSPEED_ZERO_X) && defined(DRAGSPEED_ZERO_Y)
+  dragspeed.zero_calibrated = TRUE;
+#else
+  dragspeed.zero_calibrated = FALSE;
+#endif
+  // Register callbacks
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_DRAGSPEED,
+      send_dragspeed);
+  AbiBindMsgIMU_ACCEL_INT32(DRAGSPEED_ACCEL_ID, &accel_ev, accel_cb);
+}
+
+bool dragspeed_calibrate_coeff(void)
+{
+  dragspeed.calibrate_coeff = TRUE;
+  return FALSE;
+}
+
+bool dragspeed_calibrate_zero(void)
+{
+  dragspeed.calibrate_zero = TRUE;
+  return FALSE;
+}
+
+bool dragspeed_is_calibrating(void)
+{
+  return dragspeed.calibrate_coeff || dragspeed.calibrate_zero;
+}
+
+static void accel_cb(uint8_t sender_id __attribute__((unused)), uint32_t stamp,
+    struct Int32Vect3 *accel)
+{
+  // Estimate current velocity
+  float vx = -(ACCEL_FLOAT_OF_BFP(accel->x) - dragspeed.zero.x)
+      / dragspeed.coeff.x;
+  float vy = -(ACCEL_FLOAT_OF_BFP(accel->y) - dragspeed.zero.y)
+      / dragspeed.coeff.y;
+  // Simple low-pass filter
+  dragspeed.vel.x += (1 - dragspeed.filter) * (vx - dragspeed.vel.x);
+  dragspeed.vel.y += (1 - dragspeed.filter) * (vy - dragspeed.vel.y);
```

Comment:
Not something you need to change but something you should keep in mind it that implementing the low pass filter like this assumes that you have a fixed update of the acceleros. This assumption is probably invariant while flying but is not invariant across vehicles e.g. bebop has an update rate of 500Hz but ardrone has 200Hz so this filter will be different on those vehicle for the same value of dragspeed.filter.

Like I said this is not something that you have to fix but should keep in mind.